Better Search Workflow (IOS-141)

zeitschlag opened the pull request:
We improved the Search Flow and were inspired by [the amazing Ivory](https://tapbots.com/ivory/).

Search History looks like this now:

<img src="https://github.com/mastodon/mastodon-ios/assets/2580019/a36e8101-6cf8-4200-bcc1-78561a04deb1" width="300px">

It features a condensed version of the `UserView`.

When you search for things, there are two groups now. The upper one offers some handy shortcuts, while the lower group suggests up to three hashtags and up to three users:

<img src="https://github.com/mastodon/mastodon-ios/assets/2580019/7c029199-2914-4d57-9feb-20fd5fc1f96e" width="300px">

If you paste a link into the hashtag, another elements becomes visible:

<img src="https://github.com/mastodon/mastodon-ios/assets/2580019/819a423b-48f6-495d-8c52-7815cdf54ab3" width="300px">

When this links leads to a status or user, you can see them directly inside the app. Oh, and we added some error-handling, too. Let's say, the user you're looking for doesn't exist, then alert tells you so:

<img src="https://github.com/mastodon/mastodon-ios/assets/2580019/a2269bc0-bfc9-4cfc-98af-fe1bbf3bab63" width="300px">

But wait, that's not all! You can search for users on other servers, too. Just enter their handle...

<img src="https://github.com/mastodon/mastodon-ios/assets/2580019/9971042b-f92e-4708-a5db-78b305472f93" width="300px">

... and the "Go to..."-option shows the way to them right away :)

zeitschlag committed:
Remove some logs while figuring out how search works (IOS-141)

zeitschlag committed:
More cleanup (IOS-141)

zeitschlag committed:
Show title instead of "x" for clear-all-button (IOS-141)

Also: Move stuff around. It doesn't make sense for search-diffable-stuff to live in an enitrely different folder than the rest.

zeitschlag committed:
WIP: Add some search-implementation and clean stuff (IOS-141)

Shame on me for such a big commit. I'm new to iOS-development, sorry :nerd:

zeitschlag committed:
Show up to three hashtags and up to three users (IOS-141)

zeitschlag committed:
Show hashtag-page (IOS-141)

zeitschlag committed:
Show profile-page for suggested accounts in search (IOS-141)

zeitschlag committed:
Search for people (IOS-141)

zeitschlag committed:
Show posts with a certain keyword instead of Hashtag (IOS-141)

zeitschlag committed:
Sprinkle in some localization (IOS-141)

zeitschlag committed:
Add recent searches to recent searches (IOS-141)

zeitschlag committed:
Go to user (IOS-141)

zeitschlag committed:
Add localization to "No such user"-alert (IOS-141)

zeitschlag committed:
Show max. 10 items in history (IOS-141)

zeitschlag committed:
Seach-key dismisses keyboard (IOS-141)

zeitschlag committed:
Go to link (IOS-141)

It's either a profile (then show the profile) or a status (than show the status) or a link (then open the link in safari)

zeitschlag committed:
Consider mastodon-domains for other instances and profiles (IOS-141)

zeitschlag committed:
Fix build (IOS-141)

zeitschlag committed:
Show magnifying glass next to hashtags (IOS-141)

zeitschlag committed:
Don't show follow-button for me (IOS-141)

zeitschlag committed:
Add a cell for profiles in search results (IOS-141)

zeitschlag committed:
Copy condensed version of user-view to collection-view in search-history (IOS-141)

zeitschlag committed:
Put CondensedUserView into their own class (IOS-141)

zeitschlag committed:
Don't prioritize hashtags or users (IOS-141)

zeitschlag committed:
Fix separator-inset (IOS-141)

zeitschlag committed:
Fix separator-inset on search-history (IOS-141)

zeitschlag committed:
Add https if there's no prefix (IOS-141)

zeitschlag committed:
Refactor navigation-logic into a coordinator (IOS-141)

zeitschlag committed:
Fix background-color (IOS-141)

zeitschlag committed:
Adopt search to follow mastodon-logic (IOS-141)

as defined here: https://github.com/mastodon/mastodon/blob/main/app/javascript/mastodon/features/compose/components/search.jsx#L280-L310

zeitschlag committed:
Only search for accounts when searching for a person (IOS-141)

zeitschlag committed:
Add Hashtag-Regex (IOS-141)

DK1DA commented on Mastodon/Protocol/Provider/DataSourceFacade+Profile.swift:
```diff
@@ -127,227 +127,6 @@ extension DataSourceFacade {
         let barButtonItem: UIBarButtonItem?
     }
     
-//    @MainActor
```
🙌

DK1DA commented on Mastodon/Protocol/Provider/DataSourceFacade+SearchHistory.swift:
```diff
@@ -8,11 +8,12 @@
     
     static func responseToCreateSearchHistory(
-        provider: DataSourceProvider & AuthContextProvider,
+        provider: NeedsDependency & UIViewController & AuthContextProvider,
```
We're using `NeedsDependency & UIViewController` quite often, maybe it makes sense to create a `typealias` for this?

zeitschlag replied:
it's.... 🥁 🥁 🥁 

🎉 `ViewControllerWithDependencies` 🎉

DK1DA commented on Mastodon/Scene/Search/SearchDetail/SearchDetailViewController.swift:
```diff
@@ -23,22 +22,20 @@ final class CustomSearchController: UISearchController {
 
 // Fake search bar not works on iPad with UISplitViewController
 // check device and fallback to standard UISearchController
-final class SearchDetailViewController: PageboyViewController, NeedsDependency {
```
Yes, I like that we got rid of `Pageboy`here. 👏

kimar reviewed (approved):
Looks really great @zeitschlag just one tiny thing which, when using "Posts matching XYZ" (or People matching XYZ) there is a certain delay between showing the view and when the activity indicator appears. In the first place to me this looked like there were no results and I just saw an empty screen. Maybe we can show the activity indicator immediate once the screen is show but has no results yet?

zeitschlag committed:
Show loading-indicator (IOS-141)

aaaaand simplify things as we don't need a super-dynamic search-result-screen anymore.

zeitschlag committed:
Add a typealias for `NeedsDependency & UIViewController` (IOS-141)

Well, okay, `ViewControllerWithDependencies` not super creative.
